pkg/azure/api/api.go

```diff
@@ -310,6 +354,14 @@ func generateIpConfigName() string {
 // AssignPrivateIpAddresses assigns the IPs to the interface as specified by
 // the interfaceName.
 func (c *Client) AssignPrivateIpAddresses(ctx context.Context, instanceID, vmssName, subnetID, interfaceName string, addresses int) error {
+	if vmssName == "" {
```

Comment:
I'm not sure this is the right place to fork the code and do the `vm` vs. `vmss` logic. `AssignPrivateIpAddresses` gets called via https://github.com/cilium/cilium/blob/master/pkg/azure/ipam/node.go#L141. 

How about we change the internal Azure API shim to be:
- `AssignPrivateIpAddressesVM`
- `AssignPrivateIpAddressesVMSS` -- this one following the implementation on the master branch currently

`AssignPrivateIpAddressesVM` -- to be called in https://github.com/cilium/cilium/blob/master/pkg/azure/ipam/node.go#L135 based on the `vmss` being not `""` in the `iface` object we cast via https://github.com/cilium/cilium/blob/master/pkg/azure/ipam/node.go#L136.

The current interface for the internal Azure shim is at https://github.com/cilium/cilium/blob/master/pkg/azure/ipam/instances.go#L29-L34.

@tgraf wdyt about this proposal?

I think another overhaul we could do is to piggy back just on the `vmID` field from https://github.com/cilium/cilium/blob/master/pkg/azure/types/types.go#L122-L123 if we set this correctly as Azure resource names. I believe that one to have the `vmss` name in it and we just have to write some extraction functions. But this is another story which we could tackle later if the code gets too confusing.

Reply:
Yes I also considered that option, and don't have much preference either way.

Chose to keep `AssignPrivateIpAddresses` to remain consistent with / comparable to [`EC2API` interface](https://github.com/cilium/cilium/blob/master/pkg/aws/eni/instances.go#L42) (eg. hints what remains to be implemented for a given cloud provider, and perhaps might be later a base for a common cloud API interface).

Not sure if keeping AWS and Azure APIs similar really matters though?
Let me know what's the preferred way.

Reply:
I think they are different and we never had the discussion to consolidate into one shim. My preference is to not do that given all the subtle differences between AWS/Azure (and also GCP for the future).

Reply:
@ungureanuvladvictor implemented your suggestion